client/allocrunner/task_coordinator.go

```diff
@@ -0,0 +1,386 @@
+package allocrunner
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// taskCoordinatorState represents a state of the task coordinatator FSM.
+type taskCoordinatorState string
+
+const (
+	taskCoordinatorStateInit      taskCoordinatorState = "init"
+	taskCoordinatorStatePrestart                       = "prestart"
+	taskCoordinatorStateMain                           = "main"
+	taskCoordinatorStatePoststart                      = "poststart"
+	taskCoordinatorStatePoststop                       = "poststop"
+)
+
+// taskLifecycle represents a lifecycle configuration of interest for the task
+// coordination. Not all possible combinations of hook x sidecar are defined,
+// only the ones that are relevant for task coordinator.
+type taskLifecycle string
+
+const (
+	// taskLifecyclePrestartEphemeral are tasks with the "prestart" hook and
+	// sidecar set to "false".
+	taskLifecyclePrestartEphemeral taskLifecycle = "prestart_ephemeral"
+
+	// taskLifecyclePrestartSidecar are tasks with the "prestart" hook and
+	// sidecar set to "true".
+	taskLifecyclePrestartSidecar = "prestart_sidecar"
+
+	// taskLifecycleMain are tasks without a lifecycle or a lifecycle with an
+	// empty hook value.
+	taskLifecycleMain = "main"
+
+	// taskLifecyclePoststart are tasks with the "poststart" hook.
+	taskLifecyclePoststart = "poststart"
+
+	// taskLifecyclePoststop are tasks with the "poststop" hook.
+	taskLifecyclePoststop = "poststop"
+)
+
+// taskCoordinator controls when tasks with a given lifecycle configuration are
+// allowed to run.
+//
+// It behaves like a finite state machine where each state transition blocks or
+// allows some task lifecycle type to run.
+type taskCoordinator struct {
+	logger hclog.Logger
+
+	// tasksByLifecycle is an index used to group and quickly access tasks by
+	// their lifecycle configuration.
+	tasksByLifecycle map[taskLifecycle][]string
+
+	// currentState is the current state of the FSM. It must only be accessed
+	// while holding the lock.
+	currentState     taskCoordinatorState
+	currentStateLock sync.RWMutex
+
+	// controllers store a controller for each task lifecycle.
+	controllers map[taskLifecycle]*taskCoordinatorController
+}
+
+// newTaskCoordinator returns a new taskCoordinator in the "init" state and
+// with all tasks blocked.
+func newTaskCoordinator(logger hclog.Logger, tasks []*structs.Task, shutdownCh <-chan struct{}) *taskCoordinator {
```

Comment:
**_feel free to ignore_**

using channels to coordinate shutdowns is a practice from the early days of Go - nowadays it's trendy to use a Context with a Cancel func, is that something that would make sense here?

Reply:
I tried to do this in #14127, but it was a bigger change than I expected since the channels in the `TaskRunner` and `AllocRunner` are used in several places, so I had to drop it.